src/efftester.ml

```diff
@@ -104,64 +114,41 @@ type term =
   | Let of variable * etype * term * term * etype * eff
   | If of etype * term * term * term * eff
 
-
 (** Printing functions  *)
-      
-let rec typeToOCaml ?(effannot=false) sb = function
-  | Typevar a  -> Buffer.add_string sb ("'a" ^ (string_of_int a))
-  | Unit       -> Buffer.add_string sb "unit"
-  | Int        -> Buffer.add_string sb "int"
-  | Bool       -> Buffer.add_string sb "bool"
-  | String     -> Buffer.add_string sb "string"
-  | List s     ->
-    (Buffer.add_string sb "(";
-     typeToOCaml ~effannot:effannot sb s;
-     Buffer.add_string sb ") list")
-  | Fun (s, e, t) ->
+
+let rec typeToOCaml ?(effannot = false) sb = function
+  | Typevar a -> Printf.bprintf sb "'a%d" a
+  | Unit -> Buffer.add_string sb "unit"
+  | Int -> Buffer.add_string sb "int"
+  | Bool -> Buffer.add_string sb "bool"
+  | String -> Buffer.add_string sb "string"
+  | List s -> Printf.bprintf sb "(%a) list" (typeToOCaml ~effannot) s
+  | Fun (s, (ef, ev), t) ->
+    let bprintf_fun_type format_str format_str_effannot =
+      let typeToOCamlAnnot = typeToOCaml ~effannot in
+      if effannot
+      then
+        Printf.bprintf sb format_str_effannot typeToOCamlAnnot s ef ev typeToOCamlAnnot t
+      else Printf.bprintf sb format_str typeToOCamlAnnot s typeToOCamlAnnot t
+    in
     (match s with
-      | Unit 
-      | Int
-      | Bool
-      | String ->
-        (typeToOCaml ~effannot:effannot sb s;
-	 if effannot
-	 then
-	   (Buffer.add_string sb " -";
-	    Buffer.add_string sb (string_of_bool (fst e));
-            Buffer.add_string sb "/";
-	    Buffer.add_string sb (string_of_bool (snd e));
-            Buffer.add_string sb "-> ")
-	 else
-	   Buffer.add_string sb " -> ";
-         typeToOCaml ~effannot:effannot sb t)
-      | _      ->
-        (Buffer.add_string sb "(";
-         typeToOCaml ~effannot:effannot sb s;
-         Buffer.add_string sb ")";
-	 if effannot
-	 then
-	   (Buffer.add_string sb " -";
-	    Buffer.add_string sb (string_of_bool (fst e));
-            Buffer.add_string sb "/";
-	    Buffer.add_string sb (string_of_bool (snd e));
-            Buffer.add_string sb "-> ")
-	 else
-	   Buffer.add_string sb " -> ";
-         typeToOCaml ~effannot:effannot sb t))
-
-let typetostr ?(effannot=false) typ =
+    | Unit | Int | Bool | String | List _ | Typevar _ ->
+      bprintf_fun_type "%a -> %a" "%a -%B/%B-> %a"
+    | Fun _ -> bprintf_fun_type "(%a) -> %a" "(%a) -%B/%B-> %a")
```

Comment:
You are right that this code is more complex than what was there before, but this can be improved. The key is to move the conditional behavior to be as local as possible. Currently you are testing (1) whether the arrow should be effect-annotated or not and (2) whether the left-hand-type should be parenthesized or not, bot at the toplevel, which makes 4 cases to handle (corresponding to the four format strings in your code). With the right auxiliary definitions, you could write the following instead:

```ocaml
Printf.bprintf "%a -%a> %a"
  print_simple_type s
  print_effannot (if effanot then Some e else None)
  print_type t
```

A type expression is "simple" if it is clearly delimited, either because it is a single word, or because it starts and ends with matching delimiters. `print_simple_type` prints any type expression as a simple type: in the `Fun _` case, it adds parentheses around the type.

Reply:
That's a nice solution. The function application, however, looks like this 
```ocaml
Printf.bprintf
      sb
      "%a -%a> %a"
      print_simple_type
      s
      print_effannot
      (if effannot then Some e else None)
      print_type
      t
```. When the ocamlformat v0.10 is published, it will have an option `break-fun-sig`, which might solve this problem.